Added get_packages() and standardise_dates()

henriquesposito opened the pull request:
# Description

These patch changes create a function for gathering and installing other packages part of the qDatr ecosystem, `get_packages()`, a function `standardise_dates()` replaces ambiguous dates with a range of dates, and tests and fixes to the `setup_package()`, `export_data()` and `import_data()` functions were also added.

# Package 

- Fixes #61 by adding and updating tests for `setup_package()`, `import_data()` and `export_data()`  
- Closes #50 by making sure workflow templates are created in the proper folder with `setup_package()`

# Correction

- Closes #62 by creating a function which vectorizes dates and cleans and standardizes diverse date formats, `standardise_dates()` 
- Closes #55 by deprecating `rearrange()` function in favor of `dplyr::relocate()` 
Closes #65 by correcting the path call in `export_data()` function

# Connection

- Closes #64 by creating the `get_packages()` function which displays all the other packages from the qDatr ecosystem and some information about each package

# Checklist:

- [x] I have commented my code in-line, particularly in hard-to-understand areas
- [x] My code follows the style guidelines of this project
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have documented any functions or data in the corresponding .R scripts so that roxygen can create documentation
- [x] The package builds on my OS without issues
- [x] My changes generate no new warnings
- [x] PR checks all pass for latest commit
- [x] I have bumped the version by the appropriate increment in the DESCRIPTION file (major, minor, patch)
- [x] I have updated NEWS.md
- [x] I have given this pull request an informative title and description

jhollway committed:
Fixed bug on pathing

jhollway committed:
Re #61 avoided NEWS based user prompts

jhollway committed:
Added qNEWS template

jhollway committed:
Percolated pushrelease changes through to all qPackages

jhollway committed:
Fixed #50 by relying on file.copy instead of anything routed through whisker render

jhollway committed:
Renamed qData-raw to qData-prep to simplify reference

jhollway committed:
Closed #62 with hierarchy added to data-raw folder

henriquesposito committed:
Updated TODO tasks in import_data(), export_data() and setup_package()

henriquesposito committed:
Merge branch 'develop' of https://github.com/globalgov/qDatr into develop

# Conflicts:
#	man/import_data.Rd

henriquesposito committed:
Updated error description in test_setup_package.R and TODO lines for possible tests in test_export_data.R

henriquesposito committed:
Added a depricated warning for rearrange() function in favor of dplyr::relocate

The rearrange() function still works but it now has the depricated warning

jhollway committed:
Solved test_setup_package conflict

jhollway committed:
Closed #64 by adding get_packages(), which gets information on packages in the qDatr ecosystem if empty otherwise wraps remotes::install_github()

jhollway committed:
.github/workflows files only if interactive

jhollway committed:
Changed qtemplate to use pathing instead of project

jhollway committed:
Re #61 setup_package now relies on pathing, which fixes this testing issue for all native templates, but LICENSE and NAMESPACE currently still rely on usethis

jhollway committed:
Re #61 updated create_test_package() helper to stick strictly to pathing rather than project, usethis based templates not yet working

jhollway committed:
Pause rearrange() tests since deprecated

jhollway committed:
Re #61 pausing import_/export_data tests until used with new create_test_package() helper

jhollway committed:
Added ambig_date() to replace ambiguous dates with ranges

jhollway committed:
Added battery of tests for ambig_date(), future/historical dates not yet passing

jhollway committed:
Added todos on future and historical dates

henriquesposito committed:
Removed wrong false call from export_data() function

henriquesposito committed:
Merge branch 'develop' of https://github.com/globalgov/qDatr into develop

henriquesposito committed:
Added missing exclamation point to get_packages() function

jhollway commented on tests/testthat/test_ambigdate.R:
```diff
@@ -0,0 +1,59 @@
+  expect_match(as.character(max(ambig_date("2010-01-01_03")[[1]])), "2010-01-03")
+})
+
+test_that("ambig_date() treats future dates correctly",{
+  expect_match(as.character(ambig_date("9999-12-31")[[1]]), "9999-12-31")
+  expect_match(as.character(ambig_date("2599-12-31")[[1]]), "9999-12-31")
+})
```
Either fix ambig_date so that it converts (extreme) future dates to a consistent 9999-12-31 or skip this test for now

henriquesposito replied:
For now the tests were skipped, but I think this is an important issue to come back to rather soon. Should I write this as an issue for the future in the board or should this be fixed before this pull request is merged?

jhollway replied:
Yes, please add it as an issue so that we can merge soon.

jhollway commented on tests/testthat/test_ambigdate.R:
```diff
@@ -0,0 +1,59 @@
+})
+
+test_that("ambig_date() treats historical dates correctly",{
+  expect_match(as.character(ambig_date("1712-01-01")[[1]]), "1712-01-01")
+  expect_match(as.character(ambig_date("712-01-01")[[1]]), "0712-01-01")
+  expect_match(as.character(ambig_date("0712-01-01")[[1]]), "0712-01-01")
+  expect_match(as.character(min(ambig_date("712")[[1]])), "0712-01-01")
+  expect_match(as.character(min(ambig_date("712 AD")[[1]])), "0712-01-01")
+  expect_match(as.character(min(ambig_date("712 BC")[[1]])), "-0712-01-01")
+  expect_match(as.character(min(ambig_date("-712")[[1]])), "-0712-01-01")
+  expect_match(as.character(min(ambig_date("-1712")[[1]])), "-1712-01-01")
```
Either fix ambig_date() so that it converts historical dates to expected dates class or skip these tests for now

jhollway commented on inst/templates/qPackage-NEWS.md:
```diff
@@ -0,0 +1,5 @@
+# {{{package}}} 0.0.1
+
+* Set up {{{package}}} package
+  * Added `DESCRIPTION` file
```
Add more lines outlining the actions taken to set up the package

jhollway commented on R/package.R:
```diff
@@ -147,3 +167,48 @@ setup_package <- function(packageName = NULL,
+
+#' Find and download packages in the qDatr ecosystem
+#'
+#' Find and download packages in the qDatr ecosystem
+#' @param pkg A character vector of package names 
+#' @details Need to write up these details...
+#' @return If no package name is provided, this function prints a table (tibble) to the console 
+#' with details on packages that are currently available within the qDatr ecosystem.
+#' If one or more package names are provided, these will be installed from 
```
Expand on the documentation for this function

jhollway commented on R/package.R:
```diff
@@ -147,3 +167,48 @@ setup_package <- function(packageName = NULL,
+#' If one or more package names are provided, these will be installed from 
+#' @importFrom pointblank %>%
+#' @importFrom dplyr filter
+#' @importFrom stringr str_detect
```
Add importFrom for as_tibble and fromJSON

jhollway commented on R/ambigdate.R:
```diff
@@ -0,0 +1,91 @@
+#' Create nested vectors of dates from vague date inputs
+#'
+#' @param x String vector of potential dates.
+#' @return Nested vector of POSIXct dates that includes
```
Add much more documentation

henriquesposito replied:
I am having trouble documenting this function...

jhollway replied:
Take a look at the (\dontrun) example above it, as well as the tests, and think about what this function does compared to other functions: it seeks to convert ambiguous and ranged dates into a range of dates, and (soon) extend the date parsing of other packages to more historical and future dates.

jhollway commented on R/ambigdate.R:
```diff
@@ -0,0 +1,91 @@
+#' ) datestest %>% print(n = 25)}
+#' }
+#' @export
+ambig_date <- function(x){
+  
```
Reconsider function name

henriquesposito replied:
Perhaps the name should be reconsidered, I am not sure though. Here are a few (not so good) suggestions: datecube(), datev(), ambidate(), vagdate(), vdate() and, finally, qdate()...

jhollway commented on R/data.R:
```diff
@@ -2,7 +2,8 @@
 #'
 #' Create a data-raw folder and provide templates that make easier for setting up the data cleaning 
 #' and wrangling, consistent with the qDatr ecosystem
-#' @param name Intended (short)name of the dataset
+#' @param dataset Intended (short)name of the dataset
+#' @param database Intended (short)name of the dataset
```
Add more advice/documentation on what a dataset/database or data cube is in our terminology

jhollway commented on R/data.R:
```diff
@@ -17,39 +18,48 @@
+import_data <- function(dataset = NULL,
+                        database = NULL,
+                        # type = NULL,
                         path = NULL,
                         delete_original = FALSE,
                         open = rlang::is_interactive()) {
 
   # Step one: checks and setup
-  stopifnot(rlang::is_string(name)) # Could also check if ASCII
+  if(is.null(dataset)) stop("You need to name the dataset. We suggest a short name, all small letters, such as 'cow'.")
+  if(is.null(database)) stop("You need to name the database to which the dataset would belong. We suggest a descriptive short name, all small letters, such as 'states'.")
+  stopifnot(rlang::is_string(dataset)) # Could also check if ASCII
+  stopifnot(rlang::is_string(database)) # Could also check if ASCII
   usethis::use_directory("data-raw", ignore = TRUE)
-  usethis::ui_done("Made sure data-raw folder exists.") 
+  usethis::use_directory(paste("data-raw", database, sep = "/"), ignore = TRUE)
+  usethis::use_directory(paste(paste("data-raw", database, sep = "/"), dataset, sep = "/"), ignore = TRUE)
+  usethis::ui_done("Made sure data folder hierarchy exists.") 
```
Need to declare these import dependencies?

henriquesposito replied:
I do not think so, but I do think it is better to read and understand the function when they are declared. In any case, would you like me to remove them?

jhollway replied:
No, not remove: but if we keep them do we need to declare them in the documentation, e.g. `@importFrom usethis ui_done`?

jhollway commented on R/data.R:
```diff
@@ -17,39 +18,48 @@
-  if (grepl("csv$", path)) impcmd <- "readr::read_csv"
-  if (grepl("xlsx$|xls$", path)) impcmd <- "readxl::read_excel"
-  if (grepl("dta$", path)) impcmd <- "haven::read_dta"
+  if (grepl("csv$", path)) import_type <- "readr::read_csv"
+  if (grepl("xlsx$|xls$", path)) import_type <- "readxl::read_excel"
+  if (grepl("dta$", path)) import_type <- "haven::read_dta"
   # TODO: Add these packages as suggests or maybe have the function install them if necessary
```
Add these packages as suggests or maybe have the function install them if necessary, and remove this line

jhollway commented on tests/testthat/test_setup_package.R:
```diff
@@ -1,35 +1,46 @@
+  if (fs::dir_exists(dir)) {
+    usethis::ui_stop("Target {usethis::ui_code('dir')} {usethis::ui_path(dir)} already exists.")
+  }
   
-  # change working directory
-  setwd(dir) # B
-  withr::defer(setwd(old_project), envir = env) # -B
+  old_wd <- getwd() # not necessarily same as `old_project`
   
-  # switch to new usethis project
-  usethis::proj_set(dir) # C
-  withr::defer(usethis::proj_set(old_project, force = TRUE), envir = env) # -C
+  withr::defer(
+    {
+      usethis::ui_done("Deleting temporary project: {usethis::ui_path(dir)}")
+      fs::dir_delete(dir)
+    },
+    envir = env
+  )
+  usethis::ui_silence(qDatr::setup_package("qTest", "hs", path = dir)) # A
   
-  invisible(dir)
+  withr::defer(
+    {
+      ui_done("Restoring original working directory: {usethis::ui_path(old_wd)}")
+      setwd(old_wd)
+    },
+    envir = env
+  )
+
```
Decide if this is going to continue giving us issues with R CMD Check, and if so, comment out/pause/delete

jhollway commented on tests/testthat/test_setup_package.R:
```diff
@@ -1,35 +1,46 @@
+  create_test_package(dir = dir)
+  expect_true(file.exists(paste0(dir, "/DESCRIPTION")))
+  expect_true(file.exists(paste0(dir, "/NEWS.md")))
+  expect_true(file.exists(paste0(dir, "/README.Rmd")))
+  # expect_true(file.exists(paste0(dir, "/NAMESPACE")))
+  expect_true(file.exists(paste0(dir, "/.github/CODE_OF_CONDUCT.md")))
```
Create NAMESPACE and LICENSE without usethis (project) dependency

henriquesposito replied:
Would you like that setup_package() creates NAMESPACE and LICENSE without using usethis? And, if so, why? I am just making sure I understand the issue here.

jhollway replied:
Only necessary if we're going to use `create_test_package()` to test `setup_package()`

henriquesposito committed:
Created templates script for qtemplate() function

henriquesposito committed:
Placed qTemplate() function into a separate script called template

henriquesposito committed:
Added line to create data folder in export_data() function

henriquesposito committed:
Added more descriptive text to qPackage-NEWS.md template

henriquesposito committed:
Updated details for get_packages() function description

henriquesposito committed:
Updated suggested packages in description file ta match import_data() suggestions for loading raw data

henriquesposito committed:
Updated path with export_data() function to fix compression issue #65

jhollway committed:
Added tibble and jsonlite dependencies

jhollway committed:
Paused future and historical dates tests

henriquesposito committed:
Added jasonlite and tibble as import dependencies in get_packages() function

henriquesposito committed:
Paused future and historical dates tests

Merge branch 'develop' of https://github.com/globalgov/qDatr into develop

# Conflicts:
#	R/package.R

henriquesposito committed:
Documented path call for qtemplate() function on template script

henriquesposito committed:
Paused tests for package creation by commenting them out and added tests for error messages

henriquesposito committed:
Added test for warning errors for import_data() function

henriquesposito committed:
Updated documentation for qTemplate() function

henriquesposito committed:
Updated documentation on database and dataset calls for import_data() function

jhollway commented on R/data.R:
```diff
@@ -2,7 +2,12 @@
 #' and wrangling, consistent with the qDatr ecosystem
-#' @param name Intended (short)name of the dataset
+#' @param dataset Intended (short)name of the dataset. This refers to the two-diemnsional sheet of data in the 
+#' form of afrom a dataset which will be connected to others in the future.  
```
spelling. duplication.

What is "in the future"?

jhollway commented on R/data.R:
```diff
@@ -2,7 +2,12 @@
-#' @param name Intended (short)name of the dataset
+#' @param dataset Intended (short)name of the dataset. This refers to the two-diemnsional sheet of data in the 
+#' form of afrom a dataset which will be connected to others in the future.  
+#' @param database Intended (short)name of the database. This refers to a domain-issue which the dataset(s) to be 
```
Where should we define "domain-issue"?

jhollway commented on R/data.R:
```diff
@@ -2,7 +2,12 @@
+#' @param dataset Intended (short)name of the dataset. This refers to the two-diemnsional sheet of data in the 
+#' form of afrom a dataset which will be connected to others in the future.  
+#' @param database Intended (short)name of the database. This refers to a domain-issue which the dataset(s) to be 
+#' corrected and connected in the future belongs to. That is, what is the domain-issue of the several datasets
```
What is "in the future"?

jhollway commented on R/data.R:
```diff
@@ -2,7 +2,12 @@
+#' form of afrom a dataset which will be connected to others in the future.  
+#' @param database Intended (short)name of the database. This refers to a domain-issue which the dataset(s) to be 
+#' corrected and connected in the future belongs to. That is, what is the domain-issue of the several datasets
+#' to be connected belong to. By connecting two-dimensional datasets, we can forma a three-dimensional database 
```
spelling: "forma"

jhollway commented on R/data.R:
```diff
@@ -2,7 +2,12 @@
+#' @param database Intended (short)name of the database. This refers to a domain-issue which the dataset(s) to be 
+#' corrected and connected in the future belongs to. That is, what is the domain-issue of the several datasets
+#' to be connected belong to. By connecting two-dimensional datasets, we can forma a three-dimensional database 
+#' that resembles a data cube.
```
Where should we define "data cube"?

jhollway commented on R/data.R:
```diff
@@ -17,39 +22,48 @@
-import_data <- function(name = "DATASET", 
+import_data <- function(dataset = NULL,
+                        database = NULL,
+                        # type = NULL,
```
What is `type`?

jhollway commented on R/data.R:
```diff
@@ -97,79 +113,27 @@ export_data <- function(...,
-            open = FALSE)
+            open = FALSE,
+            ignore = FALSE,
+            path=getwd())
```
spaces around equals sign

jhollway commented on R/package.R:
```diff
@@ -7,7 +7,7 @@
 #' @param update A logical indicating whether existing files should be
 #' overwritten, by default TRUE.
-#' TODO: do these checks internally and smartly...
+#' @param path A string for a particular path. Defaults to the current working directory.
```
Path to...?

jhollway commented on R/package.R:
```diff
@@ -21,13 +21,23 @@
                           packageAuthor = NULL,
-                          update = TRUE) {
+                          update = TRUE,
+                          path) {
```
keep consistent: `path` and then `if(missing(path))...` or `path = getwd()` like elsewhere

jhollway reviewed (changes requested):
Keep up the good work Henrique!

jhollway commented:
@henriquesposito don't forget to fill in the checklist at the bottom of the PR description as you go. Make sure you remember to bump the version number in DESCRIPTION (did any old functionality break?) and fill in the NEWS.

henriquesposito committed:
Updated ambig_date()function documentation within ambigdate script

henriquesposito committed:
Declared missing importa dependencies in documentation for import_data() function

henriquesposito committed:
Removed type call from import_data() function

jhollway commented on R/ambigdate.R:
```diff
@@ -34,6 +37,7 @@ ambig_date <- function(x){
   dates <- stringr::str_replace_all(dates, "-00|-\\?\\?|-NA", "") # standardising ambiguities
   dates <- stringr::str_replace_all(dates, "_", ":") # standardising ranges
   # TODO: convert future dates
+  # dates <- stringr::str_replace_all(dates, "01:01:2022 - 12:12:9999", "01:01:9999")
```
Delete

henriquesposito committed:
Removed note line from ambigdate script and fixed path description on setup_package() function documentation

henriquesposito committed:
Removed double spelling from import_data() documentation and updated parameter description

henriquesposito committed:
Fixed spelling errors in import_data() documentation

henriquesposito committed:
Added necessary white spaces in export_data() function

jhollway committed:
Resolved conflicts: added create_directory() helper, added remotes dependency for get_packages(), added .data$ to internal variables, dropped ambig_date()

jhollway committed:
Added standardise_dates() function (previously ambig_dates)

henriquesposito committed:
Merge branch 'develop' of https://github.com/globalgov/qDatr into develop

henriquesposito committed:
Updated standirdise_dates() documentation

henriquesposito committed:
Updated path parameter call in setup_package() to be consistent with other functions in qDatr

henriquesposito committed:
Fixed path issue with import_data() by setting path to working directory

henriquesposito committed:
Bumped package version in the description file

henriquesposito committed:
Updated news file by describing patch changes of pull request #66

jhollway commented on NEWS.md:
```diff
@@ -1,3 +1,21 @@
+
+## Package 
+
+* Fixes #61 by adding and updating tests for `setup_package()`, `import_data()` and `export_data()`  
```
Adding or updating?

jhollway commented on NEWS.md:
```diff
@@ -1,3 +1,21 @@
+
+* Closes #62 by creating a function which vectorizes dates and cleans and standardizes diverse date formats, `standirdise_dates()` 
+* Closes #55 by deprecating `rearrange()` function in favor of `dplyr::relocate()` 
+* Closes #65 by correcting issues with the path call in export_data() function
```
Put `export_data()` in code fences

jhollway commented on NEWS.md:
```diff
@@ -1,3 +1,21 @@
+
+## Correction
+
+* Closes #62 by creating a function which vectorizes dates and cleans and standardizes diverse date formats, `standirdise_dates()` 
```
Spelling of function name important

jhollway commented on NEWS.md:
```diff
@@ -1,3 +1,21 @@
+
+## Connection
+
+* Closes #64 by creating the `get_packages()` function which displays other packages, and some information about these packages, in the qDatr ecosystem
```
And serves as a wrapper for downloading qPackages from GitHub

henriquesposito committed:
Updated news file to correct for requested changes

jhollway commented on NEWS.md:
```diff
@@ -2,19 +2,19 @@
 ## Connection
 
-* Closes #64 by creating the `get_packages()` function which displays other packages, and some information about these packages, in the qDatr ecosystem
+* Closes #64 by creating the `get_packages()` function which displays other packages, and some information about these packages, in the qDatr ecosystem. the function serves as a wrapper for downloading qPackages from GitHub. 
```
Capitalisation? 

Were there any other notices on spelling etc from the checks?

henriquesposito committed:
Removed trailing white spaces from the qData-prep template

henriquesposito committed:
Made sure lines were not too long on qData-prep template

henriquesposito committed:
Removed trailing white spaces of files and scripts

henriquesposito committed:
Removed trailing white spaces from scripts

henriquesposito committed:
Rejoined separated long line in import_data() function

henriquesposito committed:
Removed trailing white spaces from templates

henriquesposito committed:
Splitted long lines in data script

henriquesposito committed:
Man automatic updates for changes on data script

henriquesposito committed:
Removed trailing white spaces and added spaces after comas in data script

henriquesposito committed:
Removed trailing white spaces from scripts

jhollway commented on R/data.R:
```diff
@@ -2,12 +2,19 @@
 #' Create a data-raw folder and provide templates that make easier for setting up the data cleaning 
 #' and wrangling, consistent with the qDatr ecosystem
-#' @param name Intended (short)name of the dataset
+#' @param dataset Intended (short) name of the dataset. This refers to the two-diemnsional sheet of data in the 
```
Spelling

henriquesposito committed:
Fixed spelling mistakes on data script
